ConcordiumWallet/Views/AccountsView/SendFunds/SendFund/SendFundPresenter.swift

```
@@ -424,22 +485,41 @@ class SendFundPresenter: SendFundPresenterProtocol {
     }
 
-    private func hasSufficientFunds(amount: String) -> Bool {
-        guard let cost = cost else {
-            return true
+    private func hasSufficientFunds(amount: SendFundsAmount) -> Bool {
+        guard let cost = cost else { return true }
+
+        if case SendFundsAmount.none = amount {
+            return false
         }
-        return account.canTransfer(amount: GTU(displayValue: amount),
-                                   withTransferCost: cost,
-                                   onBalance: balanceType)
+
+        switch viewModel.selectedTokenType {
+        case .ccd:
+            if case let SendFundsAmount.ccd(ccd) = amount {
+                return account.canTransfer(
+                    amount: GTU(intValue: Int(ccd.intValue)),
+                    withTransferCost: cost,
+                    onBalance: balanceType
+                )
+            }
+        case let .cis2(token: token):
+            if case let SendFundsAmount.fungibleToken(amountToken) = amount {
+                return amountToken.intValue <= token.balance
+            }
+            if case SendFundsAmount.nonFungibleToken = amount {
+                return amount.intValue == token.balance
+            }
+        }
+
+        return false
     }
 
     private func buildTransferCostParameter() -> [TransferCostParameter] {
         var costParameters: [TransferCostParameter] = []
-        if transferType == .contractUpdate, case let SendFundsTokenType.cis2(token: token) = viewModel.selectedTokenType {
+        if transferType == .contractUpdate, case let SendFundsTokenSelection.cis2(token: token) = viewModel.selectedTokenType {
             let parameters = try? dependencyProvider.mobileWallet().serializeTokenTransferParameters(
                 input: .init(
                     tokenId: token.tokenId,
-                    amount: "\(viewModel.enteredAmount?.intValue ?? 0)",
+                    amount: "\(viewModel.enteredAmount.get().intValue)",
```

I understand this may be outside the scope of what we want to change, but it's a code smell to me that you need to unwrap like that. It's better than silently falling back to 0 for sure.